khal/khalendar/event.py

```diff
@@ -843,60 +844,46 @@ def create_timezone(tz, first_date=None, last_date=None):
 
     """
-    if isinstance(tz, pytz.tzinfo.StaticTzInfo):
-        return _create_timezone_static(tz)
-
     # TODO last_date = None, recurring to infinity
 
     first_date = dt.datetime.today() if not first_date else to_naive_utc(first_date)
+    # TODO: determine if there is a way to do this for BST pjk
+    # check to see if timezone is a part of dst by checking if before or after
+    # the transition it has an offset
+    if first_date.dst() == dt.timedelta(0) and first_date.replace(fold=1).dst() == dt.timedelta(0):
+        return _create_timezone_static(tz)
+
     last_date = dt.datetime.today() if not last_date else to_naive_utc(last_date)
     timezone = icalendar.Timezone()
     timezone.add('TZID', tz)
 
-    dst = {
-        one[2]: 'DST' in two.__repr__()
-        for one, two in iter(tz._tzinfos.items())
-    }
-    bst = {
-        one[2]: 'BST' in two.__repr__()
-        for one, two in iter(tz._tzinfos.items())
-    }
-
-    # looking for the first and last transition time we need to include
-    first_num, last_num = 0, len(tz._utc_transition_times) - 1
-    first_tt = tz._utc_transition_times[0]
-    last_tt = tz._utc_transition_times[-1]
-    for num, transtime in enumerate(tz._utc_transition_times):
-        if transtime > first_tt and transtime < first_date:
-            first_num = num
-            first_tt = transtime
-        if transtime < last_tt and transtime > last_date:
-            last_num = num
-            last_tt = transtime
-
+    transition_times = _get_transition_dates_dst_for_daterange(tz, first_date.replace(tzinfo=tz),
+                                                               last_date.replace(tzinfo=tz))
     timezones = {}
-    for num in range(first_num, last_num + 1):
-        name = tz._transition_info[num][2]
+    for index in range(1, len(transition_times)):
+        name = transition_times[index].tzname()
         if name in timezones:
-            ttime = tz.fromutc(tz._utc_transition_times[num]).replace(tzinfo=None)
+            ttime = transition_times[index].replace(tzinfo=None)
             if 'RDATE' in timezones[name]:
                 timezones[name]['RDATE'].dts.append(
                     icalendar.prop.vDDDTypes(ttime))
             else:
                 timezones[name].add('RDATE', ttime)
             continue
 
-        if dst[name] or bst[name]:
+        if transition_times[index].dst() != dt.timedelta(0):
             subcomp = icalendar.TimezoneDaylight()
         else:
             subcomp = icalendar.TimezoneStandard()
 
-        subcomp.add('TZNAME', tz._transition_info[num][2])
+        subcomp.add('TZNAME', name)
         subcomp.add(
             'DTSTART',
-            tz.fromutc(tz._utc_transition_times[num]).replace(tzinfo=None))
-        subcomp.add('TZOFFSETTO', tz._transition_info[num][0])
-        subcomp.add('TZOFFSETFROM', tz._transition_info[num - 1][0])
+            tz.fromutc(transition_times[index]).replace(tzinfo=None)
```

Comment:
For some reason the `TZID` is being listed as the timezone name (i.e. EDT, CET, etc.) instead of the passed in name for the ZoneInfo (i.e. "UTC", "America/New_York", etc.).